improver/regrid/landsea.py

```diff
@@ -0,0 +1,454 @@
+
+        # new version of nearest/bilinear option with/without land-sea mask
+        elif regrid_mode in (
+            "nearest-2",
+            "nearest-with-mask-2",
+            "bilinear-2",
+            "bilinear-with-mask-2",
+        ):
+            cube = self._regrid_landsea_2(cube, target_grid, regrid_mode)
+
+        cube.attributes["title"] = (
+            MANDATORY_ATTRIBUTE_DEFAULTS["title"]
+            if regridded_title is None
+            else regridded_title
+        )
+
+        return cube
+
+    def process(self, cube, target_grid, regridded_title=None):
+        """
+        Regrids cube onto spatial grid provided by target_grid
+
+        Args:
+            cube (iris.cube.Cube):
+                Cube to be regridded
+            target_grid (iris.cube.Cube):
+                Data on the target grid. If regridding with mask, this cube
+                should contain land-sea mask data to be used in adjusting land
+                and sea points after regridding.
+            regridded_title (str or None):
+                New value for the "title" attribute to be used after
+                regridding. If not set, a default value is used.
+
+        Returns:
+            iris.cube.Cube: Regridded cube with updated attributes
+        """
+        # if regridding using a land-sea mask, check this covers the source
+        # grid in the required coordinates
+        if self.REGRID_REQUIRES_LANDMASK[self.regrid_mode]:
+            if not grid_contains_cutout(self.landmask_source_grid, cube):
+                raise ValueError("Source landmask does not match input grid")
+        return self._regrid_to_target(
+            cube, target_grid, regridded_title, self.regrid_mode
+        )
+
+
+class AdjustLandSeaPoints(BasePlugin):
+    """
+    Replace data values at points where the nearest-regridding technique
+    selects a source grid-point with an opposite land-sea-mask value to the
+    target grid-point.
+    The replacement data values are selected from a vicinity of points on the
+    source-grid and the closest point of the correct mask is used.
+    Where no match is found within the vicinity, the data value is not changed.
+    """
+
+    def __init__(self, extrapolation_mode="nanmask", vicinity_radius=25000.0):
+        """
+        Initialise class
+
+        Args:
+            extrapolation_mode (str):
+                Mode to use for extrapolating data into regions
+                beyond the limits of the source_data domain.
+                Available modes are documented in
+                `iris.analysis <https://scitools.org.uk/iris/docs/latest/iris/
+                iris/analysis.html#iris.analysis.Nearest>`_
+                Defaults to "nanmask".
+            vicinity_radius (float):
+                Distance in metres to search for a sea or land point.
+        """
+        self.input_land = None
+        self.nearest_cube = None
+        self.output_land = None
+        self.output_cube = None
```

Comment:
It seems to me that the attributes `nearest_cube`, `output_cube`, `input_land`, and `output_land` should just be variables passed in/out of `correct_where_input_true `as opposed to class attributes

Reply:
Done as suggested